src/coreclr/vm/managedmdimport.cpp

```diff
@@ -613,49 +443,168 @@ MDImpl2(void, MetaDataImport::GetParentToken, mdToken tk, mdToken* ptk)
         break;
     }
 
-    if (FAILED(hr))
-    {
-        FCThrowVoid(kBadImageFormatException);
-    }
+    return hr;
 }
 FCIMPLEND
 
-MDImpl1(void, MetaDataImport::GetScopeProps, GUID* pmvid)
+MDImpl1(HRESULT, MetaDataImport::GetScopeProps, GUID* pmvid)
 {
     FCALL_CONTRACT;
 
-    HRESULT hr;
     LPCSTR szName;
-
     IMDInternalImport *_pScope = pScope;
-    hr = _pScope->GetScopeProps(&szName, pmvid);
-    if (FAILED(hr))
-    {
-        FCThrowVoid(kBadImageFormatException);
-    }
+    return _pScope->GetScopeProps(&szName, pmvid);
 }
 FCIMPLEND
 
 
-MDImpl2(void, MetaDataImport::GetMemberRefProps,
+MDImpl2(HRESULT, MetaDataImport::GetMemberRefProps,
     mdMemberRef mr,
     ConstArray* ppvSigBlob)
 {
     FCALL_CONTRACT;
 
-    HRESULT hr;
     IMDInternalImport *_pScope = pScope;
     LPCSTR szName_Ignore;
 
-    hr = _pScope->GetNameAndSigOfMemberRef(mr, (PCCOR_SIGNATURE*)&ppvSigBlob->m_array, (ULONG*)&ppvSigBlob->m_count, &szName_Ignore);
-    if (FAILED(hr))
-    {
-        FCThrowVoid(kBadImageFormatException);
-    }
+    return _pScope->GetNameAndSigOfMemberRef(mr, (PCCOR_SIGNATURE*)&ppvSigBlob->m_array, (ULONG*)&ppvSigBlob->m_count, &szName_Ignore);
 }
 FCIMPLEND
 
 #if defined(_MSC_VER) && defined(TARGET_X86)
-#pragma optimize("", on)			// restore command line optimization defaults
+#pragma optimize("", on)    // restore command line optimization defaults
 #endif
 
+struct ResultMemory final
+{
+    INT32 length;
+    INT32* alloc;
+
+    ~ResultMemory()
+    {
+        if (alloc != NULL)
+            ::free(alloc);
+    }
+
+    void AllocateManagedArray(QCall::ObjectHandleOnStack& longResult)
+    {
+        CONTRACTL
+        {
+            THROWS;
+            MODE_PREEMPTIVE;
+            PRECONDITION(alloc != NULL);
+        }
+        CONTRACTL_END;
+
+        {
+            GCX_COOP();
+            longResult.Set(AllocatePrimitiveArray(ELEMENT_TYPE_I4, length));
```

Comment:
Would it be better to split this into two FCall/QCalls: First one gets the count and returns it to managed size, managed allocates an array of the right size as necessary, second one fills the array? It would avoid the extra unmanaged allocation and copying of the stuff around that does not happen today.

Reply:
I was trying to avoid the non-trivial cost of computing the size of the array. Creating an enumerator for the Metadata is a bit rough. I was thinking of doing that but preferred the short-lived/relatively small allocation cost as opposed to creating/destroying enumerators.

Reply:
Ok, sounds reasonable.